src/modules/mc_att_control/AttitudeControl/AttitudeControlTest.cpp

```diff
@@ -33,12 +33,95 @@
 
 #include <gtest/gtest.h>
 #include <AttitudeControl.hpp>
+#include <mathlib/math/Functions.hpp>
 
 using namespace matrix;
 
 TEST(AttitudeControlTest, AllZeroCase)
 {
 	AttitudeControl attitude_control;
-	matrix::Vector3f rate_setpoint = attitude_control.update(Quatf(), Quatf(), 0.f);
+	Vector3f rate_setpoint = attitude_control.update(Quatf(), Quatf(), 0.f);
 	EXPECT_EQ(rate_setpoint, Vector3f());
 }
+
+class AttitudeControlConvergenceTest : public ::testing::Test
+{
+public:
+	AttitudeControlConvergenceTest()
+	{
+		_attitude_control.setProportionalGain(Vector3f(.5f, .6f, .3f));
+		_attitude_control.setRateLimit(Vector3f(100, 100, 100));
+	}
+
+	void checkConvergence()
+	{
+		int i; // need function scope to check how many steps
+		Vector3f rate_setpoint(1000, 1000, 1000);
+
+		for (i = 100; i > 0; i--) {
+			// run attitude control to get rate setpoints
+			const Vector3f rate_setpoint_new = _attitude_control.update(_quat_state, _quat_goal, 0.f);
+			// rotate the simulated state quaternion according to the rate setpoint
+			_quat_state = _quat_state * Quatf(AxisAnglef(rate_setpoint_new));
+
+			// expect the error and hence also the output to get smaller with each iteration
+			if (rate_setpoint_new.norm() >= rate_setpoint.norm()) {
+				break;
+			}
+
+			rate_setpoint = rate_setpoint_new;
+		}
+
+		// we need to have reached the goal attitude
+		if (!(antipodal(_quat_state) == (antipodal(_quat_goal)))) {
+			antipodal(_quat_state).print();
+			antipodal(_quat_goal).print();
+		}
+
+		EXPECT_TRUE(antipodal(_quat_state) == antipodal(_quat_goal));
+		// it shouldn't have taken longer than an iteration timeout to converge
+		EXPECT_GT(i, 0);
+	}
+
+	Quatf antipodal(const Quatf q)
```

Comment:
- Not sure if I understand that function correctly. Basically when the real part of q is positive, it returns q, when it is negative it return -q and when zero it returns [NAN, NAN, NAN, NAN]. Is it because the same attitude can be represented by two antipodal quaternions? If yes, then I'm not sure if I like the name here because it doesn't return the antipodal quaternion but the one with the positive real value (not sure what's the technical term here).
- Why do you compute that three times for each?
- Is it safe to compare float quaternions with `==` ?

Reply:
maybe a function `bool rotationsAreEqual(_quat_state, _quat_goal)` that uses "antipodal(" (but renamed) would be nicer

Reply:
> Is it because the same attitude can be represented by two antipodal quaternions?
Exactly, I realized that depending on the exact corner case it's not enough to only distinguish based on the real part because it might be zero and then has no significant sign.

The name is indeed not the best, what about `adaptAntipodal()`? I can still wrap it for the equality check.

Reply:
I renamed it `adaptAntipodal()` but didn't wrap it because I would like to rely on the gtest stream output call to show the difference of the two values if the test fails. I should definitely make that work very soon.